### ISSUE TITLE
"Exec format error" during nixos-rebuild on 17.09 channel

### ISSUE THREAD

**Issue by Moredread** [NixOS/nixpkgs#35020]
## Issue description

I currently can't build the current nixos 17.09 channel after and on commit b81861f9fafeaf712802fa047b11e73f942c5c28. That commit belongs to issue #34786.

Error messages at the end of nixos-rebuild:

```
patching /nix/store/ndlwxwyjkcdh5wp9xz59i81plqxkfwn5-extra-utils/bin/kmod...
patching /nix/store/ndlwxwyjkcdh5wp9xz59i81plqxkfwn5-extra-utils/bin/e2fsck...
patching /nix/store/ndlwxwyjkcdh5wp9xz59i81plqxkfwn5-extra-utils/bin/tune2fs...
patching /nix/store/ndlwxwyjkcdh5wp9xz59i81plqxkfwn5-extra-utils/bin/btrfs...
patching /nix/store/ndlwxwyjkcdh5wp9xz59i81plqxkfwn5-extra-utils/bin/cryptsetup...
testing patched programs...
hello world
/build/.attr-0: line 113: /nix/store/ndlwxwyjkcdh5wp9xz59i81plqxkfwn5-extra-utils/bin/udevadm: cannot execute binary file: Exec format error
builder for '/nix/store/w7pbndgwj6xr3szaifyqq9ym4y5nbjw2-extra-utils.drv' failed with exit code 126
cannot build derivation '/nix/store/99fi6dqvr6lxdvj307kzm7igz55x1j1h-udev-rules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/w9zg76cy2g81mkc30dgn3h95liixdgz5-stage-1-init.sh.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/5qqwf59ddq0nkqn0ydp8sklaamfjr72f-initrd.drv': 1 dependencies couldn't be built
```

### Steps to reproduce

I don't have a minimal working example, but my config is available at https://github.com/Moredread/nixos-config with the machine config in `machines/grable.nix`

Building that on the 17.09 fails.

## Technical details

 - system: `"86_64-linux"`
 - host os: `Linux 4.15.0, NixOS, 17.09.2905.c1d9aff56e0 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.12pre5873_b76e282d`
 - channels(root): `"nixos-17.09, nixos-unstable-18.03pre126729.2e4aded3669"`
 - channels(addy): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs

**Comment by vcunat** [comment]
That's suspicious, and it doesn't happen for me on that commit.  I'm getting a different derivation hash, probably because of some different options, but none of these should result into the error.  I suspect FS corruption.

In your place I'd try 
```
nix-store --verify-path /nix/store/4lczhf3yxjrgankyky6ydiq7wvwsaax6-systemd-234
```
or maybe on the whole nix store
```
nix-store --verify --check-contents
```

**Comment by Moredread** [comment]
Yeah, there seems to be some corruption going on, but `nix-store --repair-path` didn't work until I disabled the `nix.autoOptimiseStore` option. Maybe it links without checking the content of the .links location?

**Comment by vcunat** [comment]
`.links` are content-addressed, so I don't think the content should be checked every time.  If FS is losing data, it's hard to defend on higher levels.

**Comment by Moredread** [comment]
A `nix-store --verify --repair`  or `nix-store --repair-path` though I would expect to fix such issues.

**Comment by Moredread** [comment]
BTW checking/repairing my store without auto optimization fixed the issue, so I'm closing the issue.

**Comment by vcunat** [comment]
That auto-optimization is here: https://github.com/NixOS/nix/issues/1281

**Comment by Moredread** [comment]
@vcunat thanks :)

**Comment by gallexme** [comment]
same problem here using nixops to deploy my machine:

> rning: working around a Linux kernel bug by creating a hole of 2097152 bytes in ‘/nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/systemd-udevd’
> patching /nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/tune2fs...
> patching /nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/udevadm...
> warning: working around a Linux kernel bug by creating a hole of 2097152 bytes in ‘/nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/udevadm’
> patching /nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/v4l_id...
> warning: working around a Linux kernel bug by creating a hole of 2097152 bytes in ‘/nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/v4l_id’
> testing patched programs...
> hello world
> /tmp/nix-build-extra-utils.drv-0/.attr-0: line 96: /nix/store/brxh74wgp086n7y89zznmd1ylkcnsxws-extra-utils/bin/udevadm: cannot execute binary file: Exec format error
